### PR TITLE
[RF] Allow BinnedLikelihood to be used without requiring it to be in a RooSimultaneous

### PR DESCRIPTION
**Pull request by will-cern**
At the moment the BinnedLikelihood feature only becomes active when doing an initSimMode inside a RooAbsTestStatistic. This change, I think, will allow for RooNLLVar to be used directly on a RooRealSumPdf and still get the BinnedLikelihood behaviour.

### CODE REVIEW

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
That's a very good improvement! I think I'll add a commit to this PR such that this also works with `BatchMode()`.

**Comment by guitargeek** [comment]
@phsft-bot build

**Review by guitargeek — APPROVED** [review]
Thank you very much Will for pushing this!

Looks good to me now, however I added some commits that change the following behavior:

1. Also make the single-channel `BinnedLikelihood` work with BatchMode
2. Change your commit such that it also works with the RooProdPdf that wraps the RooRealSumPdf
3. Implement a unit test

Since I changed quite a bit, can you please take a look again yourself at this PR and tell me here on GitHub if this goes in the direction you imagined? Then I will merge it.

File: roofit/roofitcore/test/testTestStatistics.cxx [hunk 1]
```diff
@@ -477,6 +481,67 @@ TEST_P(OffsetBinTest, CrossCheck)
...
+      pdf.setAttribute("BinnedLikelihood");
+
+      // Wrap the channel pdf in a RooProdPdf to mimic HistFactory
+      RooProdPdf prodPdf{"prod_pdf", "prod_pdf", pdf};
+
+      ws.import(prodPdf);
+   }
+
+   // We have to wrap the pdf in a RooSimultaneous, otherwise the
+   // BinnedLikelihood optimization doesn't work.
```
Comment (will-cern) [inline roofit/roofitcore/test/testTestStatistics.cxx:521]: isn;'t this comment misleading at this point?

**Comment by will-cern** [comment]
This Jonas ... I've had a look and while I'm not familiar enough with the batchmode bits you are touching I agree with your changes for supporting the case of a rooprodpdf, so thanks for adding that!

File: roofit/roofitcore/test/testTestStatistics.cxx (hunk 1 above)
Comment (guitargeek) [reply roofit/roofitcore/test/testTestStatistics.cxx]: You're right, thanks for catching this

**Comment by egpbos** [comment]
Awesome! Let's make sure Wouter also hears about this one at the next RooFit dev meeting :)

**Pull request merged**